### DISCUSSION TITLE
PSRAM

### DISCUSSION THREAD

**Discussion by podaen** [pschatzmann/arduino-audio-tools#425] [Q&A] [answered]
Is psram auto in use when selected from the menu?

**Comment by pschatzmann** [comment, marked as answer]
In my understanding it is just a precondition that you can use the related API.
You can further information on the internet: e.g. https://thingpulse.com/esp32-how-to-use-psram/

**Reply by podaen** [reply]
Is the buffer been set to the dma then?

**Comment by pschatzmann** [comment]
I don't understand your question...

**Reply by podaen** [reply]
Is the buffer of the client malloc in to the psram with the api?

**Comment by pschatzmann** [comment]
Of cause not: it is expected that memory is allocated by calling ps_malloc.
This can be avoided by doing this: https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/

**Reply by podaen** [reply]
It's beter do it static over dynamic. There is room engough to do it. As you said the ps_malloc is the way to do it.. With the buffer in psram the esp runs much smoother as I tried to explane you in #208. I hope this helps to get you over the line to make your own buffer in stead of using the api.

**Comment by pschatzmann** [comment]
In my understanding you can not allocate data is psram statically...

**Reply by podaen** [reply]
I meant ps_calloc, is better for arrays. `heap_caps_malloc_extmem_enable(limit)` will do no good, because the incoming size is unknown.

**Comment by pschatzmann** [comment]
Sure, if it is under your control, but sometimes you have to use external libraries and in this case heap_caps_malloc_extmem_enable is your only option...

**Reply by podaen** [reply]
We are talking only about clients and streams, no tls because the psram can't handle encrypted data.

**Reply by podaen** [reply]
> Sure, if it is under your control, but sometimes you have to use external libraries

You are relying on the Arduino api for the incoming audio data. Some things are just designed for general data. The WiFi is working stable but it's far from finished. You still can use the WiFi client if that is what you mean.

**Comment by podaen** [comment]
heap_caps_malloc_extmem_enable is still malloc, not calloc.

**Comment by podaen** [comment]
I have figure it out where we went wrong. It's right like you discribe on your website that unfortualy the PSRAM can not be alloced in to the global variables. But that doesn't mean you can not do them in to the member list. And then see if there is psram available. And this is the answer my friend!
